src/h3/_cy/error_system.pyx

```diff
@@ -0,0 +1,235 @@
+    of the variable assignment and block indentation.
+    """
+    yield obj
+
+
+#
+# Base exception for C library error codes
+#
+class H3BaseException(Exception):
+    """ Base H3 exception class.
+
+    Concrete subclasses of this class correspond to specific
+    error codes from the C library.
+
+    Base/abstract subclasses will have `h3_error_code = None`, while
+    concrete subclasses will have `h3_error_code` equal to their associated
+    C library error code.
+    """
+    h3_error_code = None
+
+
+#
+# A few "abstract" exceptions; organizational.
+#
+with the_error(H3BaseException) as e:
+    class H3ValueError(e, ValueError): ...
+    class H3MemoryError(e, MemoryError): ...
+    class H3GridNavigationError(e, RuntimeError): ...
+
+
+#
+# Concrete exceptions
+#
+with the_error(H3BaseException) as e:
+    class H3FailedError(e): ...
+
+with the_error(H3GridNavigationError) as e:
+    class H3PentagonError(e): ...
+
+with the_error(H3MemoryError) as e:
+    class H3MemoryAllocError(e): ...
+    class H3MemoryBoundsError(e): ...
+
+with the_error(H3ValueError) as e:
+    class H3DomainError(e): ...
+    class H3LatLngDomainError(e): ...
+    class H3ResDomainError(e): ...
+    class H3CellInvalidError(e): ...
+    class H3DirEdgeInvalidError(e): ...
+    class H3UndirEdgeInvalidError(e): ...
+    class H3VertexInvalidError(e): ...
+    class H3DuplicateInputError(e): ...
+    class H3NotNeighborsError(e): ...
+    class H3ResMismatchError(e): ...
+    class H3OptionInvalidError(e): ...
+
+
+"""
+This defines a mapping between uint32_t error codes and concrete Python
+exception classes.
+Note that we intentionally omit E_SUCCESS, as it isn't an actual error.
+"""
+error_mapping = {
+    E_FAILED:              H3FailedError,
+    E_DOMAIN:              H3DomainError,
+    E_LATLNG_DOMAIN:       H3LatLngDomainError,
+    E_RES_DOMAIN:          H3ResDomainError,
+    E_CELL_INVALID:        H3CellInvalidError,
+    E_DIR_EDGE_INVALID:    H3DirEdgeInvalidError,
+    E_UNDIR_EDGE_INVALID:  H3UndirEdgeInvalidError,
+    E_VERTEX_INVALID:      H3VertexInvalidError,
+    E_PENTAGON:            H3PentagonError,
+    E_DUPLICATE_INPUT:     H3DuplicateInputError,
+    E_NOT_NEIGHBORS:       H3NotNeighborsError,
+    E_RES_MISMATCH:        H3ResMismatchError,
+    E_MEMORY:              H3MemoryAllocError,
+    E_MEMORY_BOUNDS:       H3MemoryBoundsError,
+    E_OPTION_INVALID:      H3OptionInvalidError,
+}
+
+# Go back and modify the class definitions so that each concrete exception
+# stores its associated error code.
+for code, ex in error_mapping.items():
+    ex.h3_error_code = code
+
+
+#
+# Helper functions
+#
+
+# TODO: Move the helpers to util?
+# TODO: Unclear how/where to expose these functions. cdef/cpdef?
+
+cdef code_to_exception(H3Error err):
+    if err == E_SUCCESS:
+        return None
+    elif err in error_mapping:
+        return error_mapping[err]
+    else:
+        raise UnknownH3ErrorCode(err)
```

Comment:
It seems surprising that UnknownH3ErrorCode does not seem to derive from H3BaseException

Reply:
Makes sense. I'll change it.